be/src/io/cache/whole_file_cache.cpp

```diff
@@ -60,55 +60,94 @@ Status WholeFileCache::_generate_cache_reader(size_t offset, size_t req_size) {
     RETURN_NOT_OK_STATUS_WITH_WARN(
             io::global_local_filesystem()->exists(cache_done_file, &done_file_exist),
             "Check local cache done file exist failed.");
+
+    std::promise<Status> download_st;
+    std::future<Status> future = download_st.get_future();
     if (!done_file_exist) {
-        bool cache_dir_exist = false;
-        RETURN_NOT_OK_STATUS_WITH_WARN(
-                io::global_local_filesystem()->exists(_cache_dir, &cache_dir_exist),
-                fmt::format("Check local cache dir exist failed. {}", _cache_dir.native()));
-        if (!cache_dir_exist) {
-            RETURN_NOT_OK_STATUS_WITH_WARN(
-                    io::global_local_filesystem()->create_directory(_cache_dir),
-                    fmt::format("Create local cache dir failed. {}", _cache_dir.native()));
+        ThreadPoolToken* thread_token =
+                ExecEnv::GetInstance()->get_serial_download_cache_thread_token();
+        if (thread_token != nullptr) {
+            auto st = thread_token->submit_func([this, &download_st, cache_done_file, cache_file] {
+                auto func = [this, cache_done_file, cache_file] {
+                    bool done_file_exist = false;
+                    bool cache_dir_exist = false;
+                    RETURN_NOT_OK_STATUS_WITH_WARN(
+                            io::global_local_filesystem()->exists(_cache_dir, &cache_dir_exist),
+                            fmt::format("Check local cache dir exist failed. {}",
+                                        _cache_dir.native()));
+                    if (!cache_dir_exist) {
+                        RETURN_NOT_OK_STATUS_WITH_WARN(
+                                io::global_local_filesystem()->create_directory(_cache_dir),
+                                fmt::format("Create local cache dir failed. {}",
+                                            _cache_dir.native()));
+                    } else {
+                        // Judge again whether cache_done_file exists, it is possible that the cache
+                        // is downloaded while waiting in the thread pool
+                        RETURN_NOT_OK_STATUS_WITH_WARN(io::global_local_filesystem()->exists(
+                                                               cache_done_file, &done_file_exist),
+                                                       "Check local cache done file exist failed.");
+                    }
+                    bool cache_file_exist = false;
+                    RETURN_NOT_OK_STATUS_WITH_WARN(
+                            io::global_local_filesystem()->exists(cache_file, &cache_file_exist),
+                            "Check local cache file exist failed.");
+                    if (!done_file_exist && cache_file_exist) {
```

Comment:
Don't return here. Download files after delete cache_file, and do the next work in this function.

Reply:
done